fabric-game-test-api-v1/src/main/java/net/fabricmc/fabric/mixin/gametest/server/MainMixin.java

```diff
@@ -0,0 +1,64 @@
+import com.mojang.authlib.yggdrasil.YggdrasilAuthenticationService;
+import joptsimple.OptionParser;
+import joptsimple.OptionSet;
+import joptsimple.OptionSpec;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
+
+import net.minecraft.resource.DataPackSettings;
+import net.minecraft.resource.ResourcePackManager;
+import net.minecraft.resource.ServerResourceManager;
+import net.minecraft.server.Main;
+import net.minecraft.server.dedicated.EulaReader;
+import net.minecraft.server.dedicated.ServerPropertiesLoader;
+import net.minecraft.util.UserCache;
+import net.minecraft.util.dynamic.RegistryOps;
+import net.minecraft.util.registry.DynamicRegistryManager;
+import net.minecraft.world.level.storage.LevelStorage;
+import net.minecraft.world.level.storage.LevelSummary;
+
+import net.fabricmc.fabric.impl.client.gametest.FabricGameTestHelperImpl;
+
+@Mixin(Main.class)
+public class MainMixin {
+	@Redirect(method = "main", at = @At(value = "INVOKE", target = "Lnet/minecraft/server/dedicated/EulaReader;isEulaAgreedTo()Z"))
+	private static boolean isEulaAgreedTo(EulaReader reader) {
+		return FabricGameTestHelperImpl.ENABLED;
+	}
+
+	@Inject(method = "main", cancellable = true, locals = LocalCapture.CAPTURE_FAILHARD, at = @At(value = "INVOKE", target = "Lnet/minecraft/world/level/storage/LevelStorage$Session;readLevelProperties(Lcom/mojang/serialization/DynamicOps;Lnet/minecraft/resource/DataPackSettings;)Lnet/minecraft/world/SaveProperties;"))
+	private static void main(String[] args, CallbackInfo info, OptionParser optionParser, OptionSpec optionSpec, OptionSpec optionSpec2, OptionSpec optionSpec3, OptionSpec optionSpec4, OptionSpec optionSpec5, OptionSpec optionSpec6, OptionSpec optionSpec7, OptionSpec optionSpec8, OptionSpec optionSpec9, OptionSpec optionSpec10, OptionSpec optionSpec11, OptionSpec optionSpec12, OptionSpec optionSpec13, OptionSpec optionSpec14, OptionSet optionSet, DynamicRegistryManager.Impl impl, Path path, ServerPropertiesLoader serverPropertiesLoader, Path path2, EulaReader eulaReader, File file, YggdrasilAuthenticationService yggdrasilAuthenticationService, MinecraftSessionService minecraftSessionService, GameProfileRepository gameProfileRepository, UserCache userCache, String string, LevelStorage levelStorage, LevelStorage.Session session, LevelSummary levelSummary, DataPackSettings dataPackSettings, boolean bl, ResourcePackManager resourcePackManager, DataPackSettings dataPackSettings2, CompletableFuture completableFuture, ServerResourceManager serverResourceManager, RegistryOps registryOps) {
+		if (FabricGameTestHelperImpl.ENABLED) {
+			FabricGameTestHelperImpl.runHeadlessServer(session, resourcePackManager, serverResourceManager, impl);
```

Comment:
On a side note, we should explore using custom game providers in the loader in the future for such purposes

Reply:
Moving to another game provider opens a huge rabbit hole that I dont want to go into, this is a much neater solution imo.

Reply:
Thinking about it, it might not be as much as I initally thought but loader needs quite a bit of work to allow extending the minecraft game provider from a mod and with a stable api that we could trust to depend on in fab api.

The current way is basically the same for modders but doesnt require massive amounts of work to create stable apis in loader. It will also be easier to run in a prod server. Doing it that was has little to nothing to gain, but a huge amount of added complexitly.